designer/client/components/Visualisation/Visualisation.tsx

```diff
@@ -1,98 +1,98 @@
-
-      this.setState({
-        layout: layout.pos,
-      });
-    }, 200);
-  }
+export function useVisualisation(ref) {
+  const { data } = useContext(DataContext);
+  const [layout, setLayout] = useState<Pos>();
+  const [pushedHistoryCount, setPushedHistoryCount] = useState(-1);
 
-  componentDidMount() {
-    this.scheduleLayout();
-  }
+  useEffect(() => {
+    const layout = getLayout(data, ref.current!);
+    setLayout(layout.pos);
+  }, [data]);
 
-  UNSAFE_componentWillReceiveProps() {
-    this.scheduleLayout();
+  function pushEmptyState() {
+    window.history.pushState(document.title, window.location.pathname);
+    setPushedHistoryCount((prevState) => prevState + 1);
   }
 
-  render() {
-    const {
-      data,
-      id,
-      updatedAt,
-      downloadedAt,
-      previewUrl,
-      persona,
-    } = this.props;
-    const { pages } = data;
-
-    const wrapperStyle = this.state.layout && {
-      width: this.state.layout?.width,
-      height: this.state.layout?.height,
+  useLayoutEffect(() => {
+    pushEmptyState();
+  }, []);
+
+  const alertUser = () => {
+    if (window.confirm(i18n("leaveDesigner"))) {
+      window.history.go(-pushedHistoryCount - 1);
+    } else {
+      pushEmptyState();
+    }
+  };
+
+  useEffect(() => {
+    window.addEventListener("popstate", alertUser);
+    return () => {
+      window.removeEventListener("popstate", alertUser);
     };
+  }, []);
+
+  return { layout };
+}
 
-    return (
-      <div className="visualisation">
-        <div className="visualisation__pages-wrapper">
-          <div ref={this.ref} style={wrapperStyle}>
-            {pages.map((page, index) => (
-              <Page
-                id={id}
-                key={index}
-                data={data}
-                page={page}
-                persona={persona}
-                previewUrl={previewUrl}
-                layout={this.state.layout?.nodes[index]}
-              />
-            ))}
-
-            {this.state.layout && (
-              <Lines layout={this.state.layout} data={data} persona={persona} />
-            )}
-          </div>
+export function Visualisation(props: Props) {
+  const ref = useRef(null);
+  const { layout } = useVisualisation(ref);
+  const { data } = useContext(DataContext);
+
+  const { updatedAt, downloadedAt, previewUrl, persona } = props;
+  const { pages } = data;
+
+  const wrapperStyle = layout && {
+    width: layout?.width,
+    height: layout?.height,
+  };
+
+  return (
+    <div className="visualisation">
+      <div className="visualisation__pages-wrapper">
+        <div ref={ref} style={wrapperStyle}>
+          {pages.map((page, index) => (
+            <Page
+              key={index}
+              page={page}
+              persona={persona}
+              previewUrl={previewUrl}
+              layout={layout?.nodes[index]}
+            />
+          ))}
+
+          {layout && <Lines layout={layout} data={data} persona={persona} />}
```

Comment:
Not that big of a deal and it's more of a personal preference than anything else but the above line could be re-written as:
```suggestion
          {layout && <Lines {..{layout, data, persona}} />}
```
Which, is arguably less readable but more concise 🤓